offload_data/plugin.cpp

```diff
@@ -0,0 +1,187 @@
+#include "common/plugin.hpp"
+#include "common/threadloop.hpp"
+#include "common/switchboard.hpp"
+#include "common/phonebook.hpp"
+#include "common/data_format.hpp"
+#include "common/global_module_defs.hpp"
+
+#include <iomanip>
+#include <fstream>
+#include <numeric>
+#define STB_IMAGE_WRITE_IMPLEMENTATION
+#include "common/gl_util/lib/stb_image_write.h"
+
+using namespace ILLIXR;
+
+std::string
+getenv_or(std::string var, std::string default_) {
+        if (std::getenv(var.c_str())) {
+                return {std::getenv(var.c_str())};
+        } else {
+                return default_;
+        }
+}
+
+class offload_data : public threadloop {
+public:
+	offload_data(std::string name_, phonebook* pb_)
+		: threadloop{name_, pb_}
+		, sb{pb->lookup_impl<switchboard>()}
+		, _offload_data_reader{sb->subscribe_latest<texture_pose>("texture_pose")}
+		, _seq_expect(1)
+		, _stat_processed(0)
+		, _stat_missed(0)
+		, enable_offload{bool(std::stoi(getenv_or("ILLIXR_OFFLOAD_ENABLE", "0")))}
+		// , obj_dir{getenv_or("ILLIXR_OFFLOAD_PATH", "metrics/offloaded_data/")}
+		{
+			// Remove existing file and create folder for offloading
+			system(("rm -r " + obj_dir).c_str());
+			system(("mkdir -p " + obj_dir).c_str());
+		}
+
+		virtual skip_option _p_should_skip() override {
+			auto in = _offload_data_reader->get_latest_ro();
+			if (!in || in->seq == _seq_expect-1) {
+				// No new data, sleep
+				std::this_thread::sleep_for(std::chrono::milliseconds{1});
+				return skip_option::skip_and_yield;
+			} else {
+				if (in->seq != _seq_expect) {
+					_stat_missed = in->seq - _seq_expect;
+				} else {
+					_stat_missed = 0;
+				}
+				_stat_processed++;
+				_seq_expect = in->seq+1;
+				return skip_option::run;
+			}
+		}
+
+		void _p_one_iteration() override {
+			std::cout << "Image index: " << img_idx++ << std::endl;
+			_offload_data_container.push_back(_offload_data_reader->get_latest_ro());
+		}
+
+		virtual ~offload_data() override {
+			// Write offloaded data from memory to disk
+			if (enable_offload)
+				writeDataToDisk(_offload_data_container);
+		}
+
+private:
+	const std::shared_ptr<switchboard> sb;
+	std::unique_ptr<reader_latest<texture_pose>> _offload_data_reader;
+	long long _seq_expect, _stat_processed, _stat_missed;
+	std::vector<int> _time_seq;
+	std::vector<const texture_pose*> _offload_data_container;
+
+	int percent = 0;
+	int img_idx = 0;
+	bool enable_offload;
+	bool is_success = false;
+	std::string obj_dir = "metrics/offloaded_data/";
```

Comment:
We might not need to expose the flexibility to set output path to users, as analysis scripts will consume the output data and return the final result that users are interested.

Per the discussion with Sam, the offloaded data is written to the same place (metrics/offloaded_data/) every time running, but the path can also be changed using environment variable if necessary. (see line 35)